java-checks/src/main/java/org/sonar/java/checks/FunctionalInterfacesSpecialisedCheck.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.Objects;
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.ParameterizedTypeTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeArguments;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.plugins.java.api.tree.WildcardTree;
+
+@Rule(key = "S4276")
+public class FunctionalInterfacesSpecialisedCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CLASS, Tree.Kind.NEW_CLASS, Tree.Kind.VARIABLE);
+  }
+
+  private static final String REPORT_STRING = "Refactor this code to use the more specialised Functional Interface '";
+  private static final String UNARY_OPERATOR = "UnaryOperator<";
+
+  @Override
+  public void visitNode(Tree tree) {
+    String reportString = null;
+    if (tree.is(Tree.Kind.CLASS)) {
+      ClassTree interfaceTree = (ClassTree) tree;
+      interfaceTree.superInterfaces().stream().map(FunctionalInterfacesSpecialisedCheck::matchInterface).filter(Objects::nonNull)
```

Comment:
I think it's a bit dangerous to report issues when there is more than one interfaces, but I'm not sure.
Do you think it's valuable to report 2 issues in the following example?

```
class MySupplier implements Supplier<Integer>, Function<Integer, Integer> {

  @Override
  public Integer get() {
    return null;
  }

  @Override
  public Integer apply(Integer t) {
    return null;
  }
}
```

If it's OK for you, fine, but add an example. I can't see classes with more than on interface.

Reply:
In my opinion, in case of two or more interfaces implemented by a class we should report only one issue on the interfaces, e.g. in the above case the message should be: "Refactor this code to use the more specialised Functional Interfaces 'IntSupplier, DoubleConsumer'".